File: app/utils/collections/Grid.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayFind = ____lualib.__TS__ArrayFind
local __TS__ArraySlice = ____lualib.__TS__ArraySlice
local ____exports = {}
function ____exports.newGrid(sizeX, sizeY, initialFun)
    local _isInBounds, toIndex
    function _isInBounds(x, y)
        return x >= 0 and x < sizeX and y >= 0 and y < sizeY
    end
    function toIndex(x, y)
        return y * sizeX + x
    end
    local grid = {}
    for y = 0, sizeY - 1 do
        for x = 0, sizeX - 1 do
            grid[toIndex(x, y) + 1] = initialFun(x, y)
        end
    end
    local function get(x, y)
        local _x = math.floor(x)
        local _y = math.floor(y)
        if _isInBounds(_x, _y) then
            return grid[toIndex(_x, _y) + 1]
        else
            return undefined
        end
    end
    local function set(x, y, value)
        local _x = math.floor(x)
        local _y = math.floor(y)
        if _isInBounds(_x, _y) then
            grid[toIndex(_x, _y) + 1] = value
        end
    end
    local function cpy()
        local copyGrid = ____exports.newGrid(sizeX, sizeY, initialFun)
        for y = 0, sizeY - 1 do
            for x = 0, sizeX - 1 do
                copyGrid.set(
                    x,
                    y,
                    get(x, y)
                )
            end
        end
        return copyGrid
    end
    local function contains(value, condition)
        local ____temp_0
        if condition ~= undefined then
            ____temp_0 = condition
        else
            ____temp_0 = function(v) return v == value end
        end
        local cond = ____temp_0
        return __TS__ArrayFind(
            grid,
            function(____, v) return cond(v) end
        ) ~= undefined
    end
    local function rows()
        return sizeY
    end
    local function cols()
        return sizeX
    end
    local function sliceRow(y)
        return __TS__ArraySlice(grid, y * sizeX, (y + 1) * sizeX)
    end
    local function sliceCol(x)
        local colStartIndex = toIndex(x, 0)
        local colEndIndex = toIndex(x, sizeY - 1)
        return __TS__ArraySlice(grid, colStartIndex, colEndIndex)
    end
    local function iterate(iterFun)
        for y = 0, sizeY - 1 do
            for x = 0, sizeX - 1 do
                local result = iterFun(
                    get(x, y),
                    {x = x, y = y, idx = y * sizeX + x}
                )
                if result ~= undefined then
                    return result
                end
            end
        end
    end
    return {
        get = get,
        set = set,
        cpy = cpy,
        contains = contains,
        values = function() return grid end,
        rows = rows,
        cols = cols,
        iterate = iterate,
        sliceRow = sliceRow,
        sliceCol = sliceCol
    }
end
return ____exports

File: src/utils/collections/Grid.ts
/* eslint-disable @typescript-eslint/no-unsafe-return */
import { IGrid } from "../types/collections/IGrid";

export function newGrid<T>(sizeX: number, sizeY: number, initialFun: (x: number, y: number) => T): IGrid<T> {
    const grid: T[] = [];

    for (const y of $range(0, sizeY - 1))
        for (const x of $range(0, sizeX - 1))
            grid[toIndex(x, y)] = initialFun(x, y);

    function get(x: number, y: number): T | undefined {
        const _x = math.floor(x);
        const _y = math.floor(y);
        if (_isInBounds(_x, _y)) {
            return grid[toIndex(_x, _y)];
        } else return undefined;
    }

    function set(x: number, y: number, value: T): void {
        const _x = math.floor(x);
        const _y = math.floor(y);
        if (_isInBounds(_x, _y))
            grid[toIndex(_x, _y)] = value;
    }

    function cpy() {
        const copyGrid = newGrid<T>(sizeX, sizeY, initialFun);
        for (const y of $range(0, sizeY - 1)) {
            for (const x of $range(0, sizeX - 1)) {
                copyGrid.set(x, y, get(x, y)!);
            }
        }
        return copyGrid;
    }

    function contains(value: T): boolean;
    function contains(value: T, condition?: (val: T) => boolean): boolean {
        const cond = condition != undefined ? condition : (v: T) => v == value;
        return grid.find(v => cond(v)) != undefined;

    }

    function _isInBounds(x: number, y: number): boolean {
        return x >= 0 && x < sizeX && y >= 0 && y < sizeY;
    }

    function rows(): number {
        return sizeY;
    }
    function cols(): number {
        return sizeX;
    }

    function sliceRow(y: number): T[] {
        return grid.slice(y * sizeX, (y + 1) * sizeX);
    }

    function sliceCol(x: number): T[] {
        const colStartIndex = toIndex(x, 0);
        const colEndIndex = toIndex(x, sizeY - 1);
        return grid.slice(colStartIndex, colEndIndex);
    }

    function iterate<R>(iterFun: (value: T, data: { x: number, y: number, idx: number }) => R): R {
        for (const y of $range(0, sizeY - 1))
            for (const x of $range(0, sizeX - 1)) {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                const result = iterFun(get(x, y), { x, y, idx: y * sizeX + x });
                if (result != undefined)
                    return result;
            }


    }

    function toIndex(x: number, y: number): number {
        return y * sizeX + x;
    }

    return {
        get, set, cpy, contains, values: () => grid, rows, cols, iterate, sliceRow, sliceCol
    };
}